File: algorithms/2_medium/M/02497_maximum_star_sum_of_a_graph/solution.cpp
class Solution {
public:
    static int maxStarSum(const vector<int>& vals, const vector<vector<int>>& edges, const int k_int) {
        const size_t n = vals.size();
        const size_t k = static_cast<size_t>(k_int);
        auto adj_positive_vals = vector<vector<int>>(n, vector<int>());

        for (const auto& e : edges) {
            const int u = e[0];
            const int v = e[1];

            if (vals[v] > 0) {
                adj_positive_vals[u].push_back(vals[v]);
            }

            if (vals[u] > 0) {
                adj_positive_vals[v].push_back(vals[u]);
            }
        }

        for (auto& row : adj_positive_vals) {
            ranges::sort(row, std::greater<>());
        }

        int res = numeric_limits<int>::min();

        for (size_t u = 0; u < n; u += 1) {
            const size_t end = std::min(adj_positive_vals[u].size(), k);
            const int sum = vals[u] + std::reduce(adj_positive_vals[u].cbegin(), adj_positive_vals[u].cbegin() + end, 0);
            res = std::max(res, sum);
        }

        return res;
    }
};

File: algorithms/2_medium/M/02497_maximum_star_sum_of_a_graph/solution.rs
impl Solution {
    pub fn max_star_sum(vals: Vec<i32>, edges: Vec<Vec<i32>>, k: i32) -> i32 {
        let n = vals.len();
        let k = k as usize;
        let mut adj_positive_vals: Vec<Vec<i32>> = vec![Vec::new(); n];

        for e in edges {
            let (u, v) = (e[0], e[1]);

            if vals[v as usize] > 0 {
                adj_positive_vals[u as usize].push(vals[v as usize]);
            }

            if vals[u as usize] > 0 {
                adj_positive_vals[v as usize].push(vals[u as usize]);
            }
        }

        for row in adj_positive_vals.iter_mut() {
            row.sort_unstable_by_key(|&x| std::cmp::Reverse(x));
        }

        let mut res = i32::MIN;

        for u in 0..n {
            let sum = vals[u] + adj_positive_vals[u as usize].iter().take(k).sum::<i32>();
            res = res.max(sum);
        }

        res
    }
}
